### PR TITLE
[Attention][Feature] Migrate GLM5-Next state cache to request-owned circular tail cache

### PR DESCRIPTION
**Pull request by YanpengDing**
### What this PR does / why we need it?

Refs #15665. Replace the GLM Next indexer's paged `KPoolState` with `Glm5NextTailCache`, a request-owned FP32 ring for raw keys and gate scores. Completed pooled keys remain in the BF16 indexer cache. Each request retains one fixed tail block until it is freed, instead of allocating tail history as token pages.

The implementation updates the cache specs, planner, manager lifecycle, attention metadata, block tables, slot mapping and Triton compression together:

- **Compression and storage:** two ordered Triton launches first compress from the current chunk plus retained history, then store the final ring entries with one writer per slot. Cache accesses use the tensors' actual storage strides. The planner preserves separate main and small page classes, with indexer/tail views sharing small-page backing storage and block ownership managed by `BlockPool`.
- **Request-private lifecycle:** tail blocks are never published to or shared through the prefix cache. Prefix-hit alignment includes the compression ratio so a fresh tail resumes at a complete pool boundary. `cache_blocks` accepts both `replay_boundary` and `replay_boundaries`, as well as calls without either keyword, while remaining a no-op for the private tail.
- **Scheduling and hashing:** private tail capacity is excluded from token-page block-size resolution. When filtering the tail leaves one full group, the resolver uses that group's own block size rather than EngineCore's minimum across all original groups. A full-384/tail-16 layout therefore resolves to `(384, 384)` for DCP=1.
- **Simplification:** remove unused tail fields, metadata and the ring-capacity alias, and reuse inherited spec/manager behavior where it preserves the fixed-block contract.

Two supporting fixes are included: fresh KDA gathered states use `torch.where` to discard invalid NaN/Inf values, and the conditional model loader uses a local `WeightsMapper` to filter `rot.` weights with the current `AutoWeightsLoader` API.

Includes CPU regressions, NPU operator tests and documentation for the tail-compression contract.

### Does this PR introduce _any_ user-facing change?

GLM Next indexer tail storage and lifecycle use a fixed per-request ring. No new serving option is required.

The current `capacity = compression ratio` contract covers non-speculative execution. MTP/speculative rollback retention, its configuration guards and validation are deferred to the separate MTP workstream. This PR does not establish MTP correctness; the two-group planning regression only verifies block-size resolution.

### How was this patch tested?

**CI on `8fb61b07faa095fb000f3fe077b2b7b63efedad6`:** [E2E run 34864151928](https://github.com/vllm-project/vllm-ascend/actions/runs/34864151928) completed successfully with vLLM `84030bbe3d74d99bad477a3d2e37a973ccd8865c`.

- Full pre-commit checks and DCO passed.
- CPU UT: **4538 passed, 38 skipped**. The four tail lifecycle variants, partial-hash alignment regression and full-384/tail-16 block-size regression all passed.
- All scheduled test-matrix jobs on 310P, A2 and A3, the selected upstream test job, and the final `ci-gate` passed. Skipped jobs are not counted as executed coverage; these results do not imply end-to-end GLM coverage for every serving mode.

**Basic model QA:** performed on the earlier refactor revision `f64b9022eb4c8de8c401e7f0bf8375640a953994`, before the subsequent cleanup and compatibility fixes. The environment was Ascend A3, GLM-5.3-Flash W8A8, vLLM `a97dacb7106ee49f39f3d1fc6ae1800ff724e01d`, DP2/TP4 with EP, `--quantization ascend`, eager execution and prefix caching disabled. The fixed fixture used prompt token IDs, temperature 0, seed 1024 and a 1024-token generation limit.

All 10 requests returned HTTP 200 and `finish_reason=stop`; **8/10 passed the strict answer rule**. All three 2048-token retrieval cases passed. The two strict mismatches were `H₂O` instead of `H2O`, and `1000 - 387 = 613` instead of `613`. A historical main `0979baf` run with the same loader compatibility fix and fixture scored 9/10, already exhibiting the water-formula mismatch. The other nine complete generated token sequences matched that historical record. This was not a simultaneous baseline comparison or proof of numerical equivalence.

The basic model QA has not been rerun on the current head. GSM8K, MTP, end-to-end GLM graph/prefix-cache/concurrent-workload validation and performance benchmarking remain outside the completed model validation.

### CODE REVIEW

File: vllm_ascend/ops/triton/compute_slot_mapping.py [hunk 1]
```diff
@@ -49,7 +50,10 @@ def _compute_slot_mapping_kernel(
         offsets = i + tl.arange(0, TILE_BLOCK_SIZE)
         mask = offsets < end_idx
         pos = tl.load(positions_ptr + offsets, mask=mask, other=0).to(tl.int32)
-        if TOTAL_CP_WORLD_SIZE == 1:
+        if IS_CIRCULAR:
+            block_indices = tl.full((TILE_BLOCK_SIZE,), 0, tl.int32)
+            slot_offsets = pos % block_size
```
[the remark this answers is not in the document]
Comment (YanpengDing) [reply vllm_ascend/ops/triton/compute_slot_mapping.py:55]: `block_idx_base` is recomputed inside the per-tile loop (the `tl.min` at lines 73 / 136 / 215 sits in the loop body), not carried across iterations. With `block_indices == 0` for every lane, `block_idx_base = min(0, INT32_MAX) = 0`, so `relative_block_indices = 0` on every tile — it can never go negative.

The circular contract is one ring block per request (`max_num_blocks_per_req = 1`), addressed as `block_table[req, 0] * block_size + pos % block_size`; the new e2e test asserts exactly that (`test_glm5next_tail_slot_mapping_triton.py:36`).

The suggested `i // block_size` would break it: once `i >= block_size`, `block_idx_base = i // block_size >= 1`, the window offsets fail the `< block_table_stride` (== 1) mask, the gather returns the `other=0` fill, and `slot_ids` collapse to `pos % block_size` — physical block 0, corrupting other requests' cache. It would also pass unnoticed until the first block boundary.

Separately, `min_block_size` at :211 does not exist in that kernel — it loads `block_size` from `block_sizes_ptr`; `min_block_size` is only a parameter of the host wrapper (`compute_slot_mapping.py:351`). That suggestion would not compile.

File: vllm_ascend/patch/platform/patch_kv_cache_coordinator.py [hunk 2]
```diff
@@ -188,16 +198,21 @@ def __init__(  # type: ignore[misc]
...
         self.use_eagle = use_eagle
 
     @property
     def _cache_hit_alignment_tokens(self) -> int:
+        tail_alignment = getattr(self, "tail_pool_alignment", 1)
         if self.enable_partial_hash_hits:
-            return self.hash_block_size
-        return self.scheduler_block_size or self.lcm_block_size
+            return lcm(self.hash_block_size, tail_alignment)
+        return lcm(self.scheduler_block_size or self.lcm_block_size, tail_alignment)
```
[the remark this answers is not in the document]
Comment (YanpengDing) [reply vllm_ascend/patch/platform/patch_kv_cache_coordinator.py:215]: `from math import lcm` is already at module level (line 5) and nothing rebinds it locally, so there is no NameError path. No change needed.

File: vllm_ascend/ops/triton/compute_slot_mapping.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (YanpengDing) [reply vllm_ascend/ops/triton/compute_slot_mapping.py]: `block_idx_base` is recomputed inside the per-tile loop (the `tl.min` at lines 73 / 136 / 215 sits in the loop body), not carried across iterations. With `block_indices == 0` for every lane, `block_idx_base = min(0, INT32_MAX) = 0`, so `relative_block_indices = 0` on every tile — it can never go negative.

The circular contract is one ring block per request (`max_num_blocks_per_req = 1`), addressed as `block_table[req, 0] * block_size + pos % block_size`; the new e2e test asserts exactly that (`test_glm5next_tail_slot_mapping_triton.py:36`).

The suggested `i // block_size` would break it: once `i >= block_size`, `block_idx_base = i // block_size >= 1`, the window offsets fail the `< block_table_stride` (== 1) mask, the gather returns the `other=0` fill, and `slot_ids` collapse to `pos % block_size` — physical block 0, corrupting other requests' cache. It would also pass unnoticed until the first block boundary.

Separately, `min_block_size` at :211 does not exist in that kernel — it loads `block_size` from `block_sizes_ptr`; `min_block_size` is only a parameter of the host wrapper (`compute_slot_mapping.py:351`). That suggestion would not compile.

File: vllm_ascend/ops/triton/compute_slot_mapping.py [hunk 3]
```diff
@@ -110,8 +117,8 @@ def _compute_slot_mapping_request(
         mask = offsets < end_idx
         pos = tl.load(positions_ptr + offsets, mask=mask, other=0).to(tl.int32)
         if TOTAL_CP_WORLD_SIZE == 1:
-            block_indices = pos // block_size
-            slot_offsets = pos - block_indices * block_size
+            block_indices = tl.where(is_circular, 0, pos // block_size)
+            slot_offsets = pos % block_size
```
[the remark this answers is not in the document]
Comment (YanpengDing) [reply vllm_ascend/ops/triton/compute_slot_mapping.py:121]: `block_idx_base` is recomputed inside the per-tile loop (the `tl.min` at lines 73 / 136 / 215 sits in the loop body), not carried across iterations. With `block_indices == 0` for every lane, `block_idx_base = min(0, INT32_MAX) = 0`, so `relative_block_indices = 0` on every tile — it can never go negative.

The circular contract is one ring block per request (`max_num_blocks_per_req = 1`), addressed as `block_table[req, 0] * block_size + pos % block_size`; the new e2e test asserts exactly that (`test_glm5next_tail_slot_mapping_triton.py:36`).

The suggested `i // block_size` would break it: once `i >= block_size`, `block_idx_base = i // block_size >= 1`, the window offsets fail the `< block_table_stride` (== 1) mask, the gather returns the `other=0` fill, and `slot_ids` collapse to `pos % block_size` — physical block 0, corrupting other requests' cache. It would also pass unnoticed until the first block boundary.

Separately, `min_block_size` at :211 does not exist in that kernel — it loads `block_size` from `block_sizes_ptr`; `min_block_size` is only a parameter of the host wrapper (`compute_slot_mapping.py:351`). That suggestion would not compile.

File: vllm_ascend/models/glm5next/ops/state_ops.py [hunk 4]
```diff
@@ -21,8 +21,10 @@ def gather_initial_states(
     """Read the cache rows at ``indices``, zeroing sequences that start fresh."""
     idx = indices.to(torch.int64) * has_initial_state.to(torch.int64)
     out = state.index_select(0, idx)
-    keep = has_initial_state.view([-1] + [1] * (state.dim() - 1)).to(out.dtype)
-    return out * keep
+    keep = has_initial_state.view([-1] + [1] * (state.dim() - 1)).to(torch.bool)
+    # Fresh requests must ignore stale cache values, including NaN/Inf:
+    # multiplying such values by zero would still produce NaN.
+    return torch.where(keep, out, 0)
```
Comment (Angazenn) [inline vllm_ascend/models/glm5next/ops/state_ops.py]: 1. When there would be NaN?

File: vllm_ascend/worker/v2/attn_utils.py
Comment (Angazenn) [inline vllm_ascend/worker/v2/attn_utils.py:1093]: If you wish to use this function to view tail cache, consider to rename it with a more general name, such as `_view_kvcache_strided`

Comment (YanpengDing) [reply vllm_ascend/worker/v2/attn_utils.py:1093]: Ok. I'll make this change in the next Model RunnerV2 PR.

**Review by Angazenn — APPROVED** [review]

**Review by realliujiaxu — APPROVED** [review]
